src/feature/rend/rendclient.c

```diff
@@ -1253,7 +1253,7 @@ rend_parse_service_authorization(const or_options_t *options,
 /** The given circuit is being freed. Take appropriate action if it is of
  * interest to the client subsystem. */
 void
-rend_client_circuit_cleanup(const circuit_t *circ)
+rend_client_circuit_cleanup(const circuit_t *circ, hs_circ_cleanup_type_t type)
```

Comment:
Hm. I have mixed feelings about this cleanup type thing. On one hand it's a reasonable design, on the other hand I feel like it's a backwards step from the original nice refactoring of `hs_circ_cleanup()`. So like we managed to centralize all the cleanup code into a single function (yay!), but at the same time we introduced this cleanup type which changes the behavior of the function based on when it's called. So is this really one function, or is it actually three different functions (one for each cleanup type)?

It's hard to tell to be honest, because the type gets carried down to the deepest children of the function, and IMO that's a bit confusing.

I'd appreciate some changes to this effect:
a) Improve documentation to make it clear how each different cleanup type is used in the long-term. Right now it's not really clear to me. I see many functions only working for the `cleanup_on_free` type but I don't see the other types being used (?)
b) Improve function naming to make it clear what kind of thing each function does. For example, `hs_client_circuit_cleanup()` only works during freeing, but its name is still "cleanup". 
c) Come up with invariants to make sure that each type has been done. So if `cleanup_on_free` always comes after `cleanup_on_close` is it possible to ensure that this has indeed happened (to make sure we are not skipping any cleanup steps)?

We don't need to do perfectly all the above, but I feel like this is the messiest (in terms of tech debt) part of this branch so it would be great if we made it nicer.

Reply:
Yes indeed. This is the change I wanted more to get your feedback. I was not entirely pleased and the other option I had in mind was to have 2 different functions that is "on_close" and "on_free".

Originally, I wanted to make it through the pubsub subsystem but that wasn't really possible since, at least for the close() part, we CAN NOT have it async as in we need to remove the circuit from the map immediately.

I'm thinking that we should go with (b) as in 2 functions. The `circuit_free()` code already checks with a BUG() if the circuit was closed before so that state logic should be taken care there.

Reply:
Agreed let's do it with two functions if you also like this idea.